repository.go

```diff
@@ -3,61 +3,172 @@ package git
 import (
 	"errors"
 	"fmt"
+	"os"
 
 	"gopkg.in/src-d/go-git.v4/config"
 	"gopkg.in/src-d/go-git.v4/plumbing"
 	"gopkg.in/src-d/go-git.v4/plumbing/object"
 	"gopkg.in/src-d/go-git.v4/plumbing/protocol/packp/sideband"
 	"gopkg.in/src-d/go-git.v4/plumbing/storer"
 	"gopkg.in/src-d/go-git.v4/storage/filesystem"
-	"gopkg.in/src-d/go-git.v4/storage/memory"
 
+	"srcd.works/go-billy.v1"
 	osfs "srcd.works/go-billy.v1/os"
 )
 
 var (
-	ErrObjectNotFound     = errors.New("object not found")
-	ErrInvalidReference   = errors.New("invalid reference, should be a tag or a branch")
-	ErrRepositoryNonEmpty = errors.New("repository non empty")
-	ErrRemoteNotFound     = errors.New("remote not found")
-	ErrRemoteExists       = errors.New("remote already exists")
+	ErrObjectNotFound          = errors.New("object not found")
+	ErrInvalidReference        = errors.New("invalid reference, should be a tag or a branch")
+	ErrRepositoryNotExists     = errors.New("repository not exists")
+	ErrRepositoryAlreadyExists = errors.New("repository already exists")
+	ErrRemoteNotFound          = errors.New("remote not found")
+	ErrRemoteExists            = errors.New("remote already exists")
+	ErrWorktreeNotProvided     = errors.New("worktree should be provided")
+	ErrIsBareRepository        = errors.New("worktree not available in a bare repository")
 )
 
 // Repository giturl string, auth common.AuthMethod repository struct
 type Repository struct {
-	r map[string]*Remote
-	s Storer
+	r  map[string]*Remote
+	s  Storer
+	wt billy.Filesystem
 
 	// Progress is where the human readable information sent by the server is
 	// stored, if nil nothing is stored and the capability (if supported)
 	// no-progress, is sent to the server to avoid send this information
 	Progress sideband.Progress
 }
 
-// NewMemoryRepository creates a new repository, backed by a memory.Storage
-func NewMemoryRepository() *Repository {
-	r, _ := NewRepository(memory.NewStorage())
-	return r
+// Init create an empty git repository, based on the given Storer and worktree.
+// The worktree Filesystem is optional, if nil a bare repository is created. If
+// the given storer is not empty ErrRepositoryAlreadyExists is returned
+func Init(s Storer, worktree billy.Filesystem) (*Repository, error) {
+	r := newRepository(s, worktree)
+	_, err := r.Reference(plumbing.HEAD, false)
+	switch err {
+	case plumbing.ErrReferenceNotFound:
+	case nil:
```

Comment:
`git init` can be run in existing repositories. It _reinitializes_ then. This reinitialization is safe and does not overwrite existing info (e.g. it will create the plumbing.HEAD reference if it does not exist).